File: care/emr/models/observation.py
from django.db import models

from care.emr.models import EMRBaseModel


class Observation(EMRBaseModel):
    status = models.CharField(max_length=255)
    is_group = models.BooleanField(default=False)
    category = models.JSONField(default=dict)
    main_code = models.JSONField(default=dict)
    alternate_coding = models.JSONField(default=list)
    subject_type = models.CharField(max_length=255)
    subject_id = models.UUIDField()
    patient = models.ForeignKey("emr.Patient", on_delete=models.CASCADE)
    encounter = models.ForeignKey("emr.Encounter", on_delete=models.CASCADE)
    effective_datetime = models.DateTimeField()
    data_entered_by = models.ForeignKey(
        "users.User", on_delete=models.CASCADE, related_name="observations_entered"
    )
    performer = models.JSONField(default=dict)
    value_type = models.CharField(max_length=255)
    value = models.JSONField()
    note = models.TextField()
    body_site = models.JSONField(default=dict)
    method = models.JSONField(default=dict)
    reference_range = models.JSONField(default=list)
    interpretation = models.CharField(max_length=255)
    parent = models.UUIDField(null=True)
    questionnaire_response = models.ForeignKey(
        "emr.QuestionnaireResponse", on_delete=models.CASCADE, null=True
    )
    component = models.JSONField(default=list)

    # TODO Add index for patient -> encounter -> system -> code

File: frontend/src/types/emr/observation/observation.ts
import { Code } from "@/types/base/code/code";
import { Interpretation } from "@/types/base/qualifiedRange/qualifiedRange";
import { ObservationDefinitionReadSpec } from "@/types/emr/observationDefinition/observationDefinition";
import { QuestionType } from "@/types/questionnaire/question";
import { SubjectType } from "@/types/questionnaire/questionnaire";
import { UserReadMinimal } from "@/types/user/user";

export enum ObservationStatus {
  FINAL = "final",
  AMENDED = "amended",
  ENTERED_IN_ERROR = "entered_in_error",
}

export enum PerformerType {
  RELATED_PERSON = "related_person",
  USER = "user",
}

export interface Performer {
  type: PerformerType;
  id: string;
}

export interface ObservationReferenceRange {
  min?: number;
  max?: number;
  interpretation?: Interpretation;
}

export type QuestionnaireSubmitResultValue = {
  value?: string | null;
  unit?: Code;
  coding?: Code;
};

// Based on backend Component
export interface ObservationComponent {
  value: QuestionnaireSubmitResultValue;
  interpretation?: Interpretation;
  reference_range?: ObservationReferenceRange[];
  code?: Code | null;
  note?: string;
}

export interface CodeableConcept {
  id?: string;
  coding?: Code[];
  text?: string | null;
}

// Based on backend BaseObservationSpec
export interface ObservationBase {
  id: string; // UUID4 | null
  status: ObservationStatus;
  category?: Code | null;
  main_code?: Code | null;
  alternate_coding?: CodeableConcept | null;
  subject_type: SubjectType;
  encounter: string | null; // UUID4 | null
  effective_datetime: string; // datetime
  performer?: Performer | null;
  value_type: QuestionType;
  value: QuestionnaireSubmitResultValue;
  note?: string | null;
  body_site?: Code | null; // ValueSetBoundCoding<...>
  method?: Code | null; // ValueSetBoundCoding<...>
  reference_range?: ObservationReferenceRange[];
  interpretation?: Interpretation;
  parent?: string | null; // UUID4 | null
  questionnaire_response?: string | null; // UUID4 | null
  component?: ObservationComponent[];
}

export interface ObservationListRead extends ObservationBase {
  created_by: UserReadMinimal;
  updated_by: UserReadMinimal;
  data_entered_by?: UserReadMinimal | null;
}

export interface ObservationRead extends ObservationListRead {
  observation_definition?: ObservationDefinitionReadSpec | null;
}

export type ObservationCreate = Omit<ObservationBase, "id">;

export type ObservationUpsert = Omit<
  ObservationBase,
  "id" | "encounter" | "subject_type"
>;

export interface ObservationAnalyzeRequest {
  codes: Code[];
  page_size?: number;
}

export interface ObservationAnalyzeGroup {
  code: Code;
  results: ObservationListRead[];
}
export interface ObservationAnalyzeResponse {
  results: ObservationAnalyzeGroup[];
}

export interface ObservationUpsertRequest {
  observation: ObservationUpsert;
  observation_id?: string | null;
  observation_definition?: string | null;
}

export interface ObservationBatchUpsertRequest {
  observations: ObservationUpsertRequest[];
}

export interface ObservationBatchUpsertResponse {
  message: string;
}

export type ObservationPlotConfig = {
  id: string;
  name: string;
  groups: {
    title: string;
    codes: Code[];
  }[];
}[];
